runtime/gc/int-inf.c

```diff
@@ -124,63 +124,137 @@ void initIntInfRes (GC_state s, __mpz_struct *res,
 }
 
 /*
- * Given an __mpz_struct pointer which reflects the answer, set
- * gcState.frontier and return the answer.
- * If the answer fits in a fixnum, we return that, with the frontier
- * rolled back.
- * If the answer doesn't need all of the space allocated, we adjust
- * the sequence size and roll the frontier slightly back.
+ * Mimics the result initialization for two simultaneous results, but returns the
+ * allocated sequence that will house the results
  */
-objptr finiIntInfRes (GC_state s, __mpz_struct *res, size_t bytes) {
+GC_objptr_sequence initIntInfRes_2 (GC_state s,
+                                    __mpz_struct *lres, __mpz_struct *rres,
+                                    ARG_USED_FOR_ASSERT size_t lbytes,
+                                    ARG_USED_FOR_ASSERT size_t rbytes) {
+  GC_objptr_sequence seq = allocate_objptr_seq (s, GC_INTINF_VECTOR_HEADER, 2);
+
   GC_intInf bp;
-  int size;
+  size_t nl_limbs, nr_limbs;  // number of limbs that each result could need
+
+  assert (lbytes + rbytes <= (size_t)(s->limitPlusSlop - s->frontier));
+  // ensure that this object only looks at area after the left object
+  bp = (GC_intInf)(s->frontier + lbytes);
+  /* Amount of space for first result is amount required by the size, no more.
+   * Will need additional guaranteed heap space for the second result.
+   */
+  nl_limbs = lbytes / sizeof(mp_limb_t);
+  /* Amount of space for second result limbs is the amount of space
+   * from the end of the first until the end of the heap, since we don't need
+   * further checking for another result.
+   */
+  nr_limbs = ((size_t)(s->limitPlusSlop - (pointer)bp->obj.limbs)) / sizeof(mp_limb_t);
+  /* The _mp_alloc field is declared as int.
+   * Avoid overflowing assignments, which could happen with huge
+   * heaps.
+   */
+  lres->_mp_alloc = (int)(min(nl_limbs,(size_t)INT_MAX));
+  rres->_mp_alloc = (int)(min(nr_limbs,(size_t)INT_MAX));
+  lres->_mp_d = (mp_limb_t*)((GC_intInf)s->frontier)->obj.limbs;
+  rres->_mp_d = (mp_limb_t*)bp->obj.limbs;
+  /* is this necessary? */
+  lres->_mp_size = 0;
+  rres->_mp_size = 0;
```

Comment:
I believe so; it tells GMP that 0 limbs of the allocated limbs are actually in use.

Reply:
I just copied this comment from preexisting code for the single argument case; I'll remove both comments.